### ISSUE TITLE
Strange inconsistency in treatment of i_ controls 

### ISSUE THREAD

**Issue by muellmusik** [supercollider/supercollider#2191]
Is this intentional?
```
SynthDef(\testI, {|i_freq = 440|
	Out.ar(0, SinOsc.ar(i_freq, mul:0.1) * Env.perc.kr(2))
}).add

Synth(\testI, [\freq, 660]) // freq 440 as expected

Synth(\testI, [\i_freq, 660]) // freq 600 as expected

Pbind(\instrument, \testI, \freq, 660).play // freq 660!?

Pbind(\instrument, \testI, \i_freq, 660).play // freq 440!!?
```

I guess I could imagine the idea, but at the very least it seems inconsistent.

**Comment by jamshark70** [comment]
It is expected, and documented (somewhere). See the construction of the msgFunc in SynthDesc.

**Comment by telephon** [comment]
seems rather surprising

**Comment by muellmusik** [comment]
Principle of most surprise?

**Comment by telephon** [comment]
"One marvels about something that lies before us and is not immediately clear and understandable. Something that lies before us and is not immediately clear and understandable is a problem. All human knowledge begins with the experience of problems" (Marco Sgarbi).

**Comment by telephon** [comment]
I just remember that James McCartney said at some point that the underscore convention hadn't been a very good idea.

**Comment by snappizz** [comment]
It's a pretty crappy design. We've had at least one incident on the mailing list of someone unintentionally prefixing an argument with t_.

**Comment by jamshark70** [comment]
I think this is one of those cases where it's easy to find fault, but maybe harder to come up with better alternatives.

The rationale for the current behavior in Events (to look in the event for `freq` instead of `i_freq`) is that the rate signifier doesn't change the meaning of the parameter. In the Event (so the thinking goes), you should be concerned with the parameter's meaning rather than its implementation. Yes, there's plenty of room to disagree -- it certainly does introduce the confusion of writing `i_freq` in the SynthDef and then having to write something different elsewhere to address it. But some of the responses here sound a bit like "why the hell would anybody think of this?" when in fact there is a defensible (though arguably imperfect) rationale.

Actually, what I recall James McCartney saying is that the whole mechanism of converting function arguments to controls turned out to have unintended consequences, and if he were to do it again, he might not have supported this. So then, instead of writing:

```
SynthDef(\name, { |freq = 440, amp = 0.1, out|
	...
}).add;
```

... you'd have to write:

```
SynthDef(\name, {
	var freq, amp, out;
	#freq, amp, out = Control.names([\freq, \amp, \out]).kr([440, 0.1, 0]);
	...
}).add;
```

... which is conceptually clearer (in the sense of less hidden logic), but blastedly less convenient (not to mention less readable).

Which underscores my point: it's easy to criticize but hard to design it better.

**Comment by jamshark70** [comment]
Another example "illustrating" (*reductio ad absurdum*) why we need to get rid of function argument controls, in the interest of clarity:

```
SynthDef(\name, { |freqs = #[110, 220, 330]|
	freqs.postln;
});
-> [ an OutputProxy, an OutputProxy, an OutputProxy ]

SynthDef(\name, { |freqs ([110, 220, 330])|
	freqs.postln;
});
-> an OutputProxy
```

**Comment by snappizz** [comment]
We already have a better design than i_/t_/a_: the "rates" argument to the SynthDef, which to my knowledge implements the same functionality without the argument name confusion.

**Comment by jamshark70** [comment]
Insert an argument, and you have to remember to change two places: one at the beginning of the def and the other at the end. It's trading one set of problems for another.

**Comment by snappizz** [comment]
Indeed, "rates" has its problems -- you have to build an array parallel to the argument list, which is annoying to write and maintain. We can make "rates" more convenient by allowing the user to specify an Event associating arguments to the control types:

    // current:
    SynthDef(\name, { arg foo, bar, baz; }, [nil, nil, \tr])
    
    // proposed:
    SynthDef(\name, { arg foo, bar, baz; }, (baz: \tr))

**Comment by telephon** [comment]
@jamshark70 I completely agree that it has been easy to say that it isn't very clean but there has never been proposed a good alternative.

The "rates" argument is really inconvenient, because you separate things very far apart that belong together. 


I prefer:

```
SynthDef(\name, {
   \freqs.kr([[110, 220, 330]).postln;
});
```

**Comment by muellmusik** [comment]
To be clear, I'm not complaining about these design decisions, or even Event's behaviour. It's the inconsistency I find problematic. If you want this behaviour I can understand the rationale, but then it should be universal, and probably happen at the SynthDef construction stage. Then i_ or t_ becomes a kind of annotation which is resolved immediately. Personally though, I too prefer the the Symbol.kr/ar/ir/tr approach in terms of syntax, as it's reasonably clear and less exceptional.

If you want to complain about things though, if it were up to me I'd try to get rid of the distinction between Controls and In. Between Controls, In UGens, reading from a bus, setting a bus, and mapping to a bus you have this weird cluster of things which offer almost equivalent functionality with different syntax. ar controls have made this worse. Surely that could be done more simply.
